File: R/workflows.R
#' Prepare jobs to submit to the cluster
prepare_jobs <- function() {
  keys <- switch(context,
    heudiconv = c("site", "sid", "session"),
    fmriprep = ,
    xcpd = c("subject", "site", "sid"),
    mriqc = c("subject", "session"),
    stop("Unknown context: ", context)
  )
  if (context == "heudiconv") {
    jobs_list <- list_jobs_whole_heudiconv()
    jobs_status <- list_jobs_status_heudiconv(argv$rerun >= 2)
  }
  if (context == "fmriprep") {
    jobs_list <- list_jobs_whole_fmriprep(argv$skip_session_check)
    jobs_status <- list_jobs_status_fmriprep()
  }
  if (context == "mriqc") {
    jobs_list <- list_jobs_whole_mriqc()
    jobs_status <- list_jobs_status_mriqc(argv$rerun >= 2)
  }
  if (context == "xcpd") {
    jobs_list <- list_jobs_whole_xcpd()
    jobs_status <- list_jobs_status_xcpd()
  }
  jobs_list |>
    left_join(jobs_status, by = keys) |>
    mutate(status = coalesce(status, "todo"))
}

#' Extract todo jobs based on parsed command line arguments
#'
#' @param jobs A `data.frame()` of the required jobs.
#' @returns The extracted jobs to be done.
extract_todo <- function(jobs) {
  filter_field <- function(jobs, field) {
    if (!hasName(argv, field) || all(is.na(argv[[field]]))) {
      return(jobs)
    }
    jobs <- filter(jobs, .data[[field]] %in% argv[[field]])
    if (nrow(jobs) == 0) {
      stop("Cannot found given data.")
    }
    if (all(jobs$status == "done") && !argv$force) {
      stop("Given data have been finished. Try adding `-f` if insisted.")
    }
    jobs
  }
  jobs |>
    filter_field("subject") |>
    filter_field("site") |>
    filter_field("sid") |>
    filter_field("session") |>
    filter(
      factor(status, c("todo", "incomplete", "done")) |>
        as.integer() <= argv$rerun
    )
}

#' Execute given jobs
#'
#' @param jobs A data frame of jobs to be executed.
#' @return Invisible `NULL`. The function is used for its side effects.
execute_jobs <- function(jobs) {
  num_jobs <- nrow(jobs)
  if (num_jobs > 0) {
    if (argv$dry_run) {
      message(
        str_glue(
          "There are {num_jobs} jobs to be commited.",
          "As follows:",
          .sep = " "
        )
      )
      options(pillar.print_max = Inf)
      print(jobs)
    } else {
      withr::with_environment(
        rlang::as_environment(argv[-1]),
        rlang::exec(paste0("commit_", context), jobs)
      )
    }
  } else {
    message("All jobs are done! No jobs were commited.")
  }
  invisible()
}

File: preproc/workflows.py
"""
Workflows translated from R: prepare_jobs, extract_todo, execute_jobs.
"""
from .constants import *
from .utils import *
from .logutil import get_logger
from .fmriprep import list_jobs_whole_fmriprep, list_jobs_status_fmriprep, commit_fmriprep
from .heudiconv import list_jobs_status_heudiconv, list_jobs_whole_heudiconv
from .mriqc import list_jobs_whole_mriqc, list_jobs_status_mriqc
from .xcpd import list_jobs_whole_xcpd, list_jobs_status_xcpd

import pandas as pd


def prepare_jobs(context: str, argv: argparse.Namespace):
    keys_map = {
        "heudiconv": ["site", "sid", "session"],
        "fmriprep": ["subject", "site", "sid"],
        "xcpd": ["subject", "site", "sid"],
        "mriqc": ["subject", "session"],
    }
    keys = keys_map[context]
    if context == "heudiconv":
        jobs_list = list_jobs_whole_heudiconv()
        jobs_status = list_jobs_status_heudiconv(argv.rerun >= 2)
    if context == "fmriprep":
        jobs_list = list_jobs_whole_fmriprep(argv.skip_session_check)
        jobs_status = list_jobs_status_fmriprep()
    if context == "mriqc":
        jobs_list = list_jobs_whole_mriqc()
        jobs_status = list_jobs_status_mriqc(argv.rerun >= 2)
    if context == "xcpd":
        jobs_list = list_jobs_whole_xcpd()
        jobs_status = list_jobs_status_xcpd()
    df = pd.merge(jobs_list, jobs_status, how="left", on=keys)
    df["status"] = df["status"].fillna("todo")
    return df


def extract_todo(context: str, jobs: pd.DataFrame, argv: argparse.Namespace) -> pd.DataFrame:
    def filter_field(df: pd.DataFrame, field: str):
        val = getattr(argv, field, None)
        if not val:
            return df
        df2 = df[df[field].isin(val if isinstance(val, (list, tuple)) else [val])]
        if df2.shape[0] == 0:
            raise ValueError("No suitable data found based on your specification.")
        if all(df2.status == "done") and argv.rerun < 3:
            raise ValueError("Given data have been finished. Try adding `--rerun all` if insisted.")
        return df2
    df = jobs.copy()
    for field in ["subject", "site", "sid", "session"]:
        df = filter_field(df, field)
    # filter by rerun
    status_rank = {"todo": 1, "incomplete": 2, "done": 3}
    df = df[df["status"].map(status_rank).fillna(1) <= argv.rerun]
    return df


def execute_jobs(context: str, jobs: pd.DataFrame, argv: argparse.Namespace):
    num_jobs = jobs.shape[0]
    logger = get_logger(__name__)
    if num_jobs > 0:
        if argv.dry_run:
            logger.info(f"There are {num_jobs} jobs to be committed. As follows:")
            pd.set_option('display.max_rows', None)
            logger.info('\n' + jobs.to_string(index=False))
        else:
            # Inject environment vars
            env = args_to_env(argv)
            if context == 'fmriprep':
                from .fmriprep import commit_fmriprep
                return commit_fmriprep(jobs, **env)
            if context == 'heudiconv':
                from .heudiconv import commit_heudiconv
                return commit_heudiconv(jobs, **env)
            if context == 'mriqc':
                from .mriqc import commit_mriqc
                return commit_mriqc(jobs, **env)
            if context == 'xcpd':
                from .xcpd import commit_xcpd
                return commit_xcpd(jobs, **env)
    else:
        logger.info("All jobs are done! No jobs were committed.")
    return None
